### PR TITLE
#57443 - introduce CosmosDB communication project

### PR DESCRIPTION
**Pull request by mszymeshop**
### What
Introduction of a project, which is going to be moved to a separate package later, that is going to take care of CosmosDB communication for us.

### Why
Captain Hook is going to start relying on CosmosDB shortly and we need this capability.

### Remarks
The usage of this code is going to come in the following PRs.

### CODE REVIEW

File: src/CaptainHook.Database/CaptainHook.Database.csproj [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.2</TargetFramework>
+    <DebugType>Full</DebugType>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Autofac" Version="4.9.2" />
+    <PackageReference Include="Eshopworld.Telemetry" Version="2.1.5" />
+    <PackageReference Include="Microsoft.Azure.DocumentDB.Core" Version="2.6.0" />
```
Comment (VaclavK) [inline src/CaptainHook.Database/CaptainHook.Database.csproj]: this is quite old.. can we refresh this?

Comment (VaclavK) [reply src/CaptainHook.Database/CaptainHook.Database.csproj]: and whatever else?

File: src/CaptainHook.Database/CosmosDB/CosmosDbRepository.cs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+                    return true;
+                }
+                catch (DocumentClientException ex)
+                {
+                    if (ex.StatusCode == HttpStatusCode.NotFound && IsErrorMessageForDocumentResource(ex.Error.Message))
+                        return false;
+
+                    throw ;
+                }
```
Comment (tabman83) [inline src/CaptainHook.Database/CosmosDB/CosmosDbRepository.cs]: can use the `catch when` syntax there?

File: src/CaptainHook.Database/CosmosDB/ICosmosDbRepository.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    /// </summary>
+    public interface ICosmosDbRepository
+    {
+        /// <summary>
+        /// Defines instance of a client for cosmos db
+        /// </summary>
+        IDocumentClient DbClient { get; }
+
+        /// <summary>
+        /// Defines the collection (and from which databsae) that should be used by the repository.
```
Comment (VaclavK) [inline src/CaptainHook.Database/CosmosDB/ICosmosDbRepository.cs]: typo

File: src/CaptainHook.Database/CosmosDB/CosmosDbRepository.cs [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        private async Task<TResult> ExecuteFunction<TResult>(Func<Task<TResult>> func)
+        {
+            while (true)
+                try
+                {
+                    return await func();
+                }
+                catch (DocumentClientException ex) when (ex.StatusCode == HttpStatusCode.PreconditionFailed)
+                {
+                    throw new StaleDataException();
```
Comment (VaclavK) [inline src/CaptainHook.Database/CosmosDB/CosmosDbRepository.cs]: I do not particularly like this pattern of signalling by exception.. if you are returning some result object.. could we simply attach an error flag etc. and the reason?

File: src/CaptainHook.Database/CaptainHook.Database.csproj (hunk 1 above)
Comment (mszymeshop) [reply src/CaptainHook.Database/CaptainHook.Database.csproj]: will do

File: src/CaptainHook.Database/CosmosDB/ICosmosDbRepository.cs (hunk 3 above)
Comment (mszymeshop) [reply src/CaptainHook.Database/CosmosDB/ICosmosDbRepository.cs]: Fixed

File: src/CaptainHook.Database/CaptainHook.Database.csproj (hunk 1 above)
Comment (mszymeshop) [reply src/CaptainHook.Database/CaptainHook.Database.csproj]: Dependencies updated, however I kept them consolidated with the other dependencies in the solution. We can do another swipe when we migrate to .NET Core 3.1

File: src/CaptainHook.Database/CosmosDB/ICosmosDbRepository.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        /// <param name="data">New version of the document data</param>
+        /// <returns></returns>
+        public static Task<Document> ReplaceAsync<T>(this ICosmosDbRepository repo, string id, T data)
+        {
+            return repo.ReplaceAsync (id, data, null);
+        }
+    }
+
+    /// <summary>
+    /// Defines basic ineterface for cosmos db-based repository
```
Comment (VaclavK) [inline src/CaptainHook.Database/CosmosDB/ICosmosDbRepository.cs]: Found one more typo

File: src/CaptainHook.Database/CosmosDB/CosmosDbRepository.cs (hunk 4 above)
Comment (mszymeshop) [reply src/CaptainHook.Database/CosmosDB/CosmosDbRepository.cs]: I have now added an enabler to address this: ADO #57823

File: src/CaptainHook.Database/CosmosDB/ICosmosDbRepository.cs (hunk 5 above)
Comment (mszymeshop) [reply src/CaptainHook.Database/CosmosDB/ICosmosDbRepository.cs]: Fixed as well, but well spot :100:

File: src/CaptainHook.Database/CosmosDB/CosmosDbRepository.cs (hunk 2 above)
Comment (mszymeshop) [reply src/CaptainHook.Database/CosmosDB/CosmosDbRepository.cs]: @tabman83 I will refrain from changing this now, however this is a valid remark. We are going to rebuild exception handling as part of ADO #57823 (see @VaclavK comment), so this is likely to be addressed then.

**Pull request merged**